jaxopt/_src/gauss_newton.py

```diff
@@ -0,0 +1,170 @@
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""GaussNewton algorithm in JAX."""
+
+from typing import Any
+from typing import Callable
+from typing import NamedTuple
+from typing import Optional
+
+from dataclasses import dataclass
+
+import jax
+import jax.numpy as jnp
+
+from jaxopt._src import base
+from jaxopt._src import linear_solve
+from jaxopt.tree_util import tree_l2_norm, tree_sub
+
+
+class GaussNewtonState(NamedTuple):
+  """Named tuple containing state information."""
+  iter_num: int
+  residual: Any
+  delta: Any
+  error: float
+  gradient: Any
+  jtj: Optional[Any] = None
+  aux: Optional[Any] = None
+
+@dataclass(eq=False)
+class GaussNewton(base.IterativeSolver):
+  """Gauss-Newton nonlinear least-squares solver.
+
+  This solver finds the optimal parameters via a least-squares optimization
+
+    Given the residual function f(x): R^m -> R^n, `gauss_newton` finds a
+    local minimum of the cost function F(x):
+
+    ```
+    argmin_x F(x) = 0.5 * sum(f_i(x)**2), i = 0, ..., m - 1
+    f(x) = func(x, *args, **kwargs)
+    ```
+
+  Attributes:
+    residual_fun: a smooth function of the form ``fun(x, *args, **kwargs)``.
+    maxiter: maximum number of iterations.
+    tol: tolerance.
+    check_bracket: whether to check correctness of the bracketing interval.
+      If True, the method ``run`` cannot be jitted.
+    implicit_diff_solve: the linear system solver to use.
+    verbose: whether to print error on every iteration or not.
+      Warning: verbose=True will automatically disable jit.
+    jit: whether to JIT-compile the bisection loop (default: "auto").
+    unroll: whether to unroll the bisection loop (default: "auto").
+  """
+  residual_fun: Callable
+  maxiter: int = 30
+  tol: float = 1e-5
+  verbose: bool = False
+  implicit_diff_solve: Optional[Callable] = None
+  has_aux: bool = False
+  jit: base.AutoOrBoolean = "auto"
+  unroll: base.AutoOrBoolean = "auto"
+
+  def init_state(self,
+                 init_params: Any,
+                 *args,
+                 **kwargs) -> GaussNewtonState:
+    """Initialize the solver state.
+
+    Args:
+      init_params: pytree containing the initial parameters.
+      *args: additional positional arguments to be passed to ``fun``.
+      **kwargs: additional keyword arguments to be passed to ``fun``.
+    Returns:
+      state
+    """
+
+    # Compute actual values of state variables at init_param
+    residual, aux = self._fun_with_aux(init_params, *args, **kwargs)
+    matvec = lambda v: self._jtj_op(init_params, v, *args, **kwargs)
+    gradient = self._jt_op(init_params, residual, *args, **kwargs)
+    # Only materialize J^T.J if verbose is True for debugging
```

Comment:
I would rather remove `jtj` from the state if possible. I don't think it's a good idea that the state depends on the `verbose` option. If you really need it, I think I would rather add an option `store_jtj` in the constructor.

Reply:
I removed `jtj`. As you mentioned, this becomes more like a debugging feature that I have to think how to extract for the optimizer. Eventually, the eignevalues of the matrix might be of interest. I'll get it into a separate PR once i have an idea.